### PR TITLE
Add parallelism to e2e suite

### PR DESCRIPTION
**Pull request by awsluja**
#### Description of changes
Currently the e2e test suite takes over 3 hours to run.
We have some individual tests suites containing more than 6 tests each, so I've updated the split tests script to allow parallelism on each test job. This should save us several hours each week of waiting for e2e tests to complete; potentially thousands of hours a year.

E2E Tests results look good:
https://app.circleci.com/pipelines/github/aws-amplify/amplify-cli/11279/workflows/dadbb35c-1491-4f56-a82d-2c070385b880

Failed tests:
amplify_sudo_install_test (known issue)
schema-auth-12_pkg-w (flaky)
node-function_pkg-w (not modified/not related)
delete_pkg-l (not modified/not related)
geo-remove-3_pkg-l (flaky)

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-cli/blob/dev/CONTRIBUTING.md#tests)
- [ ] Relevant documentation is changed or added (and PR referenced)
- [ ] New AWS SDK calls or CloudFormation actions have been added to relevant test and service IAM policies

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by evcodes** [comment]
I know that the previous names for the files were auth5 or whatever, but is there a better name for these tests?

**Review by lazpavel — DISMISSED** [review]
changes LGTM, can we trigger the cloud-e2e on this?

**Comment by awsluja** [comment]
> I know that the previous names for the files were auth5 or whatever, but is there a better name for these tests?
I went with this structure so that it was easier to review initially; as it might be hard to track down the changes if I renamed all the files depending on the test

**Comment by sobolk** [comment]
> > I know that the previous names for the files were auth5 or whatever, but is there a better name for these tests?
> > I went with this structure so that it was easier to review initially; as it might be hard to track down the changes if I renamed all the files depending on the test

I think numerical suffix is fine if there's no logical grouping and we just load balance.
However `auth_5X` could be called `auth_headless_X`. So there's some room to go after those tests and get better naming.
Also sharding test cases should follow some pattern and should have only one level of nesting. I.e. `5a`, `5b` etc isn't good.

**Comment by awsluja** [comment]
> > > I know that the previous names for the files were auth5 or whatever, but is there a better name for these tests?
> > > I went with this structure so that it was easier to review initially; as it might be hard to track down the changes if I renamed all the files depending on the test
> 
> I think numerical suffix is fine if there's no logical grouping and we just load balance. However `auth_5X` could be called `auth_headless_X`. So there's some room to go after those tests and get better naming. Also sharding test cases should follow some pattern and should have only one level of nesting. I.e. `5a`, `5b` etc isn't good.

Those are all good points; I would recommend as a follow up PR, since a pure file renaming PR will be much easier to review than a rename+refactor PR

**Comment by awsluja** [comment]
> changes LGTM, can we trigger the cloud-e2e on this?

Done, I've added a link: https://app.circleci.com/pipelines/github/aws-amplify/amplify-cli/11279/workflows/dadbb35c-1491-4f56-a82d-2c070385b880

File: packages/amplify-e2e-tests/src/__tests__/schema-auth-11-b.test.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+  beforeEach(async () => {
+    projectDir = await createNewProjectDir('auth3');
+    await initJSProjectWithProfile(projectDir, {});
+  });
+
+  afterEach(async () => {
+    await deleteProject(projectDir);
+    deleteProjectDir(projectDir);
+  });
```
Comment (evcodes) [inline packages/amplify-e2e-tests/src/__tests__/schema-auth-11-b.test.ts:20]: only one test, don't need these before/after each

File: packages/amplify-e2e-tests/src/__tests__/schema-auth-11-c.test.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+  beforeEach(async () => {
+    projectDir = await createNewProjectDir('auth3');
+    await initJSProjectWithProfile(projectDir, {});
+  });
+
+  afterEach(async () => {
+    await deleteProject(projectDir);
+    deleteProjectDir(projectDir);
+  });
+
```
Comment (evcodes) [inline packages/amplify-e2e-tests/src/__tests__/schema-auth-11-c.test.ts:21]: same here

File: packages/amplify-e2e-tests/src/__tests__/schema-auth-5a.test.ts [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+  beforeEach(async () => {
+    projectDir = await createNewProjectDir('auth5a');
+    await initJSProjectWithProfile(projectDir, {});
+  });
+
+  afterEach(async () => {
+    await deleteProject(projectDir);
+    deleteProjectDir(projectDir);
+  });
```
Comment (evcodes) [inline packages/amplify-e2e-tests/src/__tests__/schema-auth-5a.test.ts:20]: same

File: packages/amplify-e2e-tests/src/__tests__/schema-auth-5b.test.ts [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+  beforeEach(async () => {
+    projectDir = await createNewProjectDir('auth5b');
+    await initJSProjectWithProfile(projectDir, {});
+  });
+
+  afterEach(async () => {
+    await deleteProject(projectDir);
+    deleteProjectDir(projectDir);
+  });
```
Comment (evcodes) [inline packages/amplify-e2e-tests/src/__tests__/schema-auth-5b.test.ts:20]: same

File: packages/amplify-e2e-tests/src/__tests__/schema-auth-5c.test.ts [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+  beforeEach(async () => {
+    projectDir = await createNewProjectDir('auth5c');
+    await initJSProjectWithProfile(projectDir, {});
+  });
+
+  afterEach(async () => {
+    await deleteProject(projectDir);
+    deleteProjectDir(projectDir);
+  });
```
Comment (evcodes) [inline packages/amplify-e2e-tests/src/__tests__/schema-auth-5c.test.ts:20]: same

File: packages/amplify-e2e-tests/src/__tests__/schema-auth-5d.test.ts [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+
+  beforeEach(async () => {
+    projectDir = await createNewProjectDir('auth5d');
+    await initJSProjectWithProfile(projectDir, {});
+  });
+
+  afterEach(async () => {
+    await deleteProject(projectDir);
+    deleteProjectDir(projectDir);
+  });
```
Comment (evcodes) [inline packages/amplify-e2e-tests/src/__tests__/schema-auth-5d.test.ts:20]: ""

File: packages/amplify-e2e-tests/src/__tests__/schema-auth-9-b.test.ts [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+  beforeEach(async () => {
+    projectDir = await createNewProjectDir('auth8');
+    await initJSProjectWithProfile(projectDir, {});
+  });
+
+  afterEach(async () => {
+    await deleteProject(projectDir);
+    deleteProjectDir(projectDir);
+  });
+
```
Comment (evcodes) [inline packages/amplify-e2e-tests/src/__tests__/schema-auth-9-b.test.ts:21]: same

File: packages/amplify-e2e-tests/src/__tests__/schema-auth-6d.test.ts [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+
+  beforeEach(async () => {
+    projectDir = await createNewProjectDir('auth6');
+    await initJSProjectWithProfile(projectDir, {});
+  });
+
+  afterEach(async () => {
+    await deleteProject(projectDir);
+    deleteProjectDir(projectDir);
+  });
```
Comment (evcodes) [inline packages/amplify-e2e-tests/src/__tests__/schema-auth-6d.test.ts:20]: same

File: packages/amplify-e2e-tests/src/__tests__/schema-auth-6c.test.ts [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+
+  beforeEach(async () => {
+    projectDir = await createNewProjectDir('auth6');
+    await initJSProjectWithProfile(projectDir, {});
+  });
+
+  afterEach(async () => {
+    await deleteProject(projectDir);
+    deleteProjectDir(projectDir);
+  });
```
Comment (evcodes) [inline packages/amplify-e2e-tests/src/__tests__/schema-auth-6c.test.ts:20]: same

File: packages/amplify-e2e-tests/src/__tests__/schema-auth-9-c.test.ts [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+
+  beforeEach(async () => {
+    projectDir = await createNewProjectDir('auth8');
+    await initJSProjectWithProfile(projectDir, {});
+  });
+
+  afterEach(async () => {
+    await deleteProject(projectDir);
+    deleteProjectDir(projectDir);
+  });
```
Comment (evcodes) [inline packages/amplify-e2e-tests/src/__tests__/schema-auth-9-c.test.ts:20]: same

File: packages/amplify-e2e-tests/src/__tests__/schema-model-a.test.ts [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+
+  beforeEach(async () => {
+    projectDir = await createNewProjectDir('model');
+    await initJSProjectWithProfile(projectDir, {});
+  });
+
+  afterEach(async () => {
+    await deleteProject(projectDir);
+    deleteProjectDir(projectDir);
+  });
```
Comment (evcodes) [inline packages/amplify-e2e-tests/src/__tests__/schema-model-a.test.ts:20]: same

File: packages/amplify-e2e-tests/src/__tests__/schema-model-b.test.ts [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+
+  beforeEach(async () => {
+    projectDir = await createNewProjectDir('model');
+    await initJSProjectWithProfile(projectDir, {});
+  });
+
+  afterEach(async () => {
+    await deleteProject(projectDir);
+    deleteProjectDir(projectDir);
+  });
```
Comment (evcodes) [inline packages/amplify-e2e-tests/src/__tests__/schema-model-b.test.ts:20]: same

File: packages/amplify-migration-tests/src/__tests__/migration_tests/overrides/api-migration-a.test.ts [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+    projRoot = await createNewProjectDir('graphql-api');
+  });
+
+  afterEach(async () => {
+    const metaFilePath = join(projRoot, 'amplify', '#current-cloud-backend', 'amplify-meta.json');
+    if (fs.existsSync(metaFilePath)) {
+      await deleteProject(projRoot, undefined, true);
+    }
+    deleteProjectDir(projRoot);
+  });
```
Comment (evcodes) [inline packages/amplify-migration-tests/src/__tests__/migration_tests/overrides/api-migration-a.test.ts:32]: just one test here too...

**Review by evcodes — CHANGES_REQUESTED** [review]
small notes that I believe existed before your PR. If you can take a look at those that would be awesome!

File: packages/amplify-e2e-tests/src/__tests__/global_sandbox-c.test.ts [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+    deleteProjectDir(projectDir);
+  });
+
+  // TODO: need a reason why we're skipping this or we should remove this test
+  it.skip('compiles schema with three models and pushes to cloud', async () => {
+    await addApiWithThreeModels(projectDir);
+    await apiGqlCompile(projectDir, true);
+    await generateModels(projectDir);
+    await amplifyPush(projectDir, true);
+  });
```
[the remark this answers is not in the document]
Comment (awsluja) [reply packages/amplify-e2e-tests/src/__tests__/global_sandbox-c.test.ts:33]: it was skipped before; I was curious, but shouldn't block this pr

File: packages/amplify-e2e-tests/src/__tests__/export-pull-d.test.ts [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+    expect(recursiveComapre(object1, object2)).toBeTruthy();
+  };
+
+  const recursiveComapre = (object1: any, object2: any): boolean => Object.keys(object1).reduce((equal, key) => {
+    if (!equal) return false;
+    if (typeof object1[key] !== 'object') {
+      return object1[key] === object2[key];
+    }
+    return recursiveComapre(object1[key], object2[key]);
+  }, true);
```
[the remark this answers is not in the document]
Comment (awsluja) [reply packages/amplify-e2e-tests/src/__tests__/export-pull-d.test.ts:56]: follow up with a refactor PR separately as prioritized

**Comment by evcodes** [comment]
I'd approve all these changes with the small fixes to be implemented in a separate PR

**Review by evcodes — APPROVED** [review]

**Review by lazpavel — APPROVED** [review]

**Pull request merged**